keps/sig-auth/3257-trust-anchor-sets/README.md

```diff
@@ -0,0 +1,1158 @@
+  using a ConfigMap that is automatically injected into every namespace.
+
+* Kubernetes webhooks and aggregated API servers are configured with a static
+  trust anchor set embedded in their config (which complicates rotation).
+
+* * At least one signer implementation uses a [cluster-scoped
+  CRD](https://cloud.google.com/traffic-director/docs/security-proxyless-setup#:~:text=Save%20the%20following%20TrustConfig%20YAML%20configuration%20to%20tell%20your%20cluster%20how%20to%20trust%20the%20issued%20certificates%3A)
+  to distribute trust anchors to workloads.
+
+Each of these approaches has downsides:
+
+* Creating a ConfigMap in every namespace requires granting over-broad
+  permissions to the responsible controller (create ConfigMaps in **all**
+  namespaces).
+
+* Static trust anchor sets embedded directly in configuration complicate the job
+  of rotating the backing CA's roots of trust.
+
+* CRDs work well, but are inappropriate for use by core APIs.
+
+Each of these approaches would be simplified by an in-tree mechanism for
+distributing trust anchor sets.
+
+### Goals
+
+* Provide a Kubernetes data type that holds X.509 trust anchors for use wherever
+  a set of X.509 trust anchors is needed.
+
+* Make it easy for certificates/v1 signer implementations to publish their
+  relevant trust anchors using a TrustAnchorSets object.
+
+* Make it easy for workloads that need access to the trust anchors that back a
+  particular certificates/v1 signer to access them within the container
+  filesystem.
+
+* Allow cluster operators to configure webhook and aggregrated API server
+  backends using that use TrustAnchorSets instead of inline PEM data.
+
+### Non-Goals
+
+* Distribute a default set of "system" trust anchors using TrustAnchorSets.
+  (This is a natural extension, though).
+
+* Handle trust anchors expressed in other forms than PEM-wrapped, DER-formatted
+  X.509.
+
+## Proposal
+
+This proposal is centered around a new cluster-scoped TrustAnchorSets resource,
+initially in the certificates/v1alpha1 API group.  The TrustAnchorSets object
+can be thought of as a specialized configmap tailored to the X.509 trust anchor
+use case.  Introducing a dedicated type allows us to attach different RBAC
+policies to TrustAnchorSets, which will typically be wide-open for reading, but
+locked-down for writing.
+
+Each TrustAnchorSets object is a container for multiple named trust anchor sets.
+Most use cases will simply use a single trust anchor set named `default`, but
+mutual TLS use cases involving multiple federating meshes will require the
+ability to express multiple, non-commingled trust anchor sets.
+
+TrustAnchorSets are scoped to a particular signer using a naming convention that
+translates the signer name into a valid Kubernetes resource name.  For example,
+the signer named `example.com/my-signer` will conventionally export any
+necessary trust anchor sets using a TrustAnchorSets object named
+`example.com-slash-my--signer`.
```

Comment:
We could also define an annotation, stronger than the convention for `metadata.name`, that records the exact signer name verbatim. The annotation could be optional but recommended.

Should we do that?

Reply:
Sure, the annotation would not be an access control measure, it'd be there to help humans.